include/nbl/video/IDescriptorPool.h

```diff
@@ -29,12 +33,156 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
+                    }
+                    else
+                    {
+                        m_linearAllocators[i] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[i]*sizeof(void*));
+                    }
+                }
+            }
 
-        uint32_t getCapacity() const { return m_maxSets; }
+            // Initialize the storages.
+            m_combinedImageSamplerStorage = std::make_unique<StorageTrivializer<SCombinedImageSampler>[]>(m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
+            m_storageImageStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE] + m_maxDescriptorCount[asset::EDT_INPUT_ATTACHMENT]);
+            m_UBO_SSBOStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_UTB_STBStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_accelerationStructureStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
+        }
+
+        ~IDescriptorPool()
+        {
+            for (auto i = 0; i < asset::EDT_COUNT; ++i)
+            {
+                if ((m_flags & IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT) && (m_maxDescriptorCount[i] > 0))
+                    _NBL_ALIGNED_FREE(m_generalAllocatorReservedSpace[i]);
+            }
+        }
+
+        // This will return the offset into the pool's descriptor storage. These offsets will be combined
+        // later with base memory addresses to get the actual memory adress where we put the core::smart_refctd_ptr<const IDescriptor>.
+        uint32_t allocateDescriptors(const asset::E_DESCRIPTOR_TYPE type, const uint32_t count)
+        {
+            const uint32_t bytesToAllocate = count * sizeof(void*);
+
+            uint32_t offset;
+            uint32_t invalidAddress;
+            if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+            {
+                offset = m_generalAllocators[type].alloc_addr(bytesToAllocate, 1u);
+                invalidAddress = core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address;
+            }
+            else
+            {
+                offset = m_linearAllocators[type].alloc_addr(bytesToAllocate, 1u);
+                invalidAddress = core::LinearAddressAllocator<uint32_t>::invalid_address;
+            }
+
+            return (offset == invalidAddress) ? ~0u : offset;
```

Comment:
you can skip the shenanigans with the `invalidAddress` (its `0xeadebeefu` for both allocators) if the address is invalid it will naturally not be in the valid `[0,descriptorCount)` address space, so you can check the return value against that

Reply:
Resolved.